File: ruby/lib/ipaddress/rle.rb
class IPAddress


  class Last
    attr_reader :val, :max_poses, :ret
    attr_writer :val

    def initialize()
      @val = nil
      @max_poses = {}
      @ret = []
    end

    def handle_last()
      if (nil == val)
        return
      end

      _last = @val
      max_rles = @max_poses[_last.part]
      if (max_rles == nil)
        max_rles = []
        @max_poses[_last.part] = max_rles
      end

      # console.log(_last.part, @max_poses)
      # puts ">>>>#{@ret}  #{@max_poses} #{max_rles}"
      max_rles.each do |idx|
        # puts "#{@ret}  #{max_rles} #{idx}"
        prev = @ret[idx]
        if (prev.cnt > _last.cnt)
          # console.log(`>>>>> last=${_last}->${idx}->prev=${prev}`)
          _last.max = false
        elsif (prev.cnt == _last.cnt)
          # nothing
        elsif (prev.cnt < _last.cnt)
          # console.log(`<<<<< last=${_last}->${idx}->prev=${prev}`)
          prev.max = false
        end
      end

      #println!("push:{}:{:?}", self.ret.len(), _last)
      max_rles.push(@ret.length)
      _last.pos = @ret.length
      @ret.push(_last)
    end
  end

  class Rle
    attr_accessor :part, :pos, :cnt, :max

    def initialize(obj)
      @part = obj[:part]
      @pos = obj[:pos]
      @cnt = obj[:cnt]
      @max = obj[:max]
    end

    def toString()
      return "<Rle@part:#{@part},pos:#{@pos},cnt:#{@cnt},max:#{@max}>"
    end

    def eq(other)
      return @part == other.part && @pos == other.pos &&
        @cnt == other.cnt && @max == other.max
    end

    def ne(other)
      return !eq(other)
    end

    def self.code(parts)
      last = Last.new()
      # println!("code")
      parts.length.times do |i|
        part = parts[i]
        # console.log(`part:${part}`)
        if (last.val && last.val.part == part)
          last.val.cnt += 1
        else
          last.handle_last()
          last.val = Rle.new({ part: part, pos: 0, cnt: 1, max: true })
        end
      end

      last.handle_last()
      return last.ret
    end
  end
end

File: go/ipaddress/rle.go
package ipaddress

import "fmt"

type Rle struct {
	Part uint16
	Pos  int
	Cnt  int
	Max  bool
}

func (r *Rle) String() string {
	return fmt.Sprintf("Rle:Part:{%d},Pos:{%d},Cnt:{%d},Max:{%d}",
		r.Part, r.Pos, r.Cnt, r.Max)
}

func (self *Rle) Equal(other Rle) bool {
	return self.Part == other.Part && self.Pos == other.Pos &&
		self.Cnt == other.Cnt && self.Max == other.Max
}

//impl<T: PartialEq> Eq for Rle<T> {}
type Last struct {
	val       *Rle
	Max_Poses map[uint16][]int
	ret       []Rle
}

func (self *Last) handle_last() {
	if self.val == nil {
		return
	}
	_last := self.val
	Max_rles, ok := self.Max_Poses[_last.Part]
	if !ok {
		Max_rles = make([]int, 0)
		self.Max_Poses[_last.Part] = Max_rles
	}
	for _, idx := range Max_rles {
		prev := &self.ret[idx]
		if prev.Cnt > _last.Cnt {
			_last.Max = false
		}
		if prev.Cnt == _last.Cnt {
			// nothing
		}
		if prev.Cnt < _last.Cnt {
			prev.Max = false
		}
	}
	self.Max_Poses[_last.Part] = append(Max_rles, len(self.ret))
	_last.Pos = len(self.ret)
	self.ret = append(self.ret, *_last)
}

func Code(Parts []uint16) []Rle {
	last := Last{nil, make(map[uint16][]int), make([]Rle, 0)}
	for i := 0; i < len(Parts); i++ {
		Part := Parts[i]
		if last.val != nil && last.val.Part == Part {
			last.val.Cnt += 1
		} else {
			last.handle_last()
			last.val = &Rle{Part, 0, 1, true}
		}
	}
	last.handle_last()
	return last.ret
}
